substrate/frame/assets/src/functions.rs

```diff
@@ -351,28 +376,35 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
 	pub(super) fn do_refund(id: T::AssetId, who: T::AccountId, allow_burn: bool) -> DispatchResult {
 		use AssetStatus::*;
 		use ExistenceReason::*;
+
 		let mut account = Account::<T, I>::get(&id, &who).ok_or(Error::<T, I>::NoDeposit)?;
 		ensure!(matches!(account.reason, Consumer | DepositHeld(..)), Error::<T, I>::NoDeposit);
 		let mut details = Asset::<T, I>::get(&id).ok_or(Error::<T, I>::Unknown)?;
 		ensure!(matches!(details.status, Live | Frozen), Error::<T, I>::IncorrectStatus);
 		ensure!(account.balance.is_zero() || allow_burn, Error::<T, I>::WouldBurn);
 
-		if let Some(deposit) = account.reason.take_deposit() {
-			T::Currency::unreserve(&who, deposit);
-		}
+		with_storage_layer(|| {
```

Comment:
@gui1117 good catch. I missed this and proposed to move checks for `ContainsHolds/Freezes` to `dead_account` function. But with the storage layers this looks more complex (or at least not common). I would rather remove those asserts from `dead_account` and place them under the block with the rest of asserts/ensure!. 
@pandres95 sorry for back and forth with this. we rarely use `with_storage_layer` at this level, and generally try to perform asserts first and do alteration later.

what you think?

Reply:
I agree the code doesn't look so nice with the storage layer everywhere.

In general, I think we should keep the storage clean for any trait we use.
Ideally also any public function.
And maybe also private functions.
Or append their name with must_rollback_on_error.
Because people don't look at doc so often I think.

Then I don't mind any implementation, maybe we can always do the checks in the caller function, or reorganize to not commit into the state before calling the function.
I don't have so much best practices.

Reply:
Lets just copy&past these checks to the functions that needs them and remove them from `dead_account`.

Reply:
Yes! I'll do that precisely. 😁

Reply:
@muharem @gui1117 it's done ([`4f50052`](https://github.com/paritytech/polkadot-sdk/pull/4530/commits/4f50052528a2605a78ea76a40582c312de9c8cee)).